src/segment-loader.js

```diff
@@ -1260,20 +1264,44 @@ export default class SegmentLoader extends videojs.EventTarget {
     // can be adjusted by the timestamp offset
     if (!segmentInfo.hasAppendedData_) {
       this.metadataQueue_.caption.push(
-        this.handleCaptions_.bind(this, simpleSegment, captions));
+        this.handleCaptions_.bind(this, simpleSegment, captionData));
       return;
     }
 
     const timestampOffset = this.sourceUpdater_.videoTimestampOffset() === null ?
       this.sourceUpdater_.audioTimestampOffset() :
       this.sourceUpdater_.videoTimestampOffset();
 
-    captions.forEach((caption) => {
-      createCaptionsTrackIfNotExists(
-        this.inbandTextTracks_, this.hls_.tech_, caption.stream);
+    const tracks = {};
+
+    // get total start/end and captions for each track/stream
+    captionData.forEach((caption) => {
+      // caption.stream is actually a track name...
+      const track = tracks[caption.stream] = tracks[caption.stream] || {
+        startTime: Infinity,
+        captions: [],
+        endTime: 0
+      };
+
+      track.startTime = Math.min(track.startTime, (caption.startTime + timestampOffset));
+      track.endTime = Math.max(track.endTime, (caption.endTime + timestampOffset));
+      track.captions.push(caption);
+    });
+    Object.keys(tracks).forEach((trackName) => {
+      const {startTime, endTime, captions} = tracks[trackName];
+      const {inbandTextTracks_} = this;
+
+      this.logger_(`adding cues from ${startTime} -> ${endTime} for ${trackName}`);
+
+      createCaptionsTrackIfNotExists(inbandTextTracks_, this.hls_.tech_, trackName);
+      // clear out any old cues. We do this because a rendition change
+      // will cause captions to be re-parsed for certain segments, as they are
+      // downloaded, but we do not want to have two of the same captions visible.
+      removeCuesFromTrack(startTime, endTime, inbandTextTracks_[trackName]);
+
       addCaptionData({
-        captionArray: [caption],
-        inbandTextTracks: this.inbandTextTracks_,
+        captionArray: captions,
```

Comment:
I originally had this forced to be an array to have consistency between using the `PartialTransmuxer` (which outputs each caption as it is seen, an object) and `FullTransmuxer` (which outputs all the captions for a segment)

Reply:
It will always be an array now due to `track[trackName].captions` being an array now.